packages/core/elasticsearch/core-elasticsearch-client-server-internal/src/agent_manager.ts

```diff
@@ -0,0 +1,188 @@
+    const { httpInstanceIndex, httpsInstanceIndex } = this.initTypeStore(type);
+
+    return (connectionOpts: ConnectionOptions): NetworkAgent => {
+      const agents = this.getAgentInstances(type, connectionOpts);
+
+      // when the factory is called for a given connection (and protocol), we use the appropriate index
+      const agentIndex = isHttps(connectionOpts) ? httpsInstanceIndex : httpInstanceIndex;
+
+      let agent = agents[agentIndex];
+
+      if (!agent) {
+        agent = this.createAgent(agentConfig, connectionOpts);
+
+        // Allow GC'ing the Agent after it has been terminated
+        const doDestroy = agent.destroy.bind(agent);
+        agent.destroy = () => {
+          doDestroy();
+          agents[agentIndex] = undefined;
+        };
+
+        // add the new Agent instance to the list of instances of that type
+        agents[agentIndex] = agent;
+      }
+
+      return agent;
+    };
+  }
+
+  private initTypeStore(type: string): { httpInstanceIndex: number; httpsInstanceIndex: number } {
+    let agentsOfType = this.agentStore[type];
+
+    if (!agentsOfType) {
+      agentsOfType = {
+        [HTTP]: [],
+        [HTTPS]: [],
+      };
+
+      this.agentStore[type] = agentsOfType;
+    } else {
+      this.logger.warn(
+        `An Agent factory for type ${type} has already been retrieved.
+        This is not optimal since the existing instance cannot be reused.
+        Please consider reusing the previously retrieved factory instead`
+      );
+    }
+
+    const httpInstanceIndex = agentsOfType[HTTP].length;
+    const httpsInstanceIndex = agentsOfType[HTTPS].length;
+
+    agentsOfType[HTTP].push(undefined);
+    agentsOfType[HTTPS].push(undefined);
+    return { httpInstanceIndex, httpsInstanceIndex };
+  }
+
+  private getAgentInstances(
+    type: string,
+    connectionOpts: ConnectionOptions
+  ): Array<NetworkAgent | undefined> {
+    let protocol: Protocol;
+
+    switch (connectionOpts.url.protocol) {
+      case 'http:':
+        protocol = HTTP;
+        break;
+      case 'https:':
+        protocol = HTTPS;
+        break;
+      default:
+        throw new Error(`Unsupported protocol: '${connectionOpts.url.protocol}'`);
+    }
+
+    return this.agentStore[type][protocol];
+  }
+
+  private createAgent(agentConfig: HttpAgentOptions, connectionOpts: ConnectionOptions) {
+    const config = Object.assign({}, this.defaultConfig, agentConfig, connectionOpts.tls);
+    return isHttps(connectionOpts) ? new HttpsAgent(config) : new HttpAgent(config);
+  }
+}
+
+const isHttps = (connectionOpts: ConnectionOptions): boolean => {
+  return connectionOpts.url.protocol === HTTPS;
+};
+
+const assertValidAgentConfig = (
+  agentOptions?: HttpAgentOptions | UndiciAgentOptions | false
+): HttpAgentOptions => {
+  if (!agentOptions) {
+    return {};
+  } else if (isHttpAgentOptions(agentOptions)) {
+    return agentOptions;
+  }
+
+  throw new Error('Unsupported agent options: UndiciAgentOptions');
```

Comment:
Are we planning (would we be able) to support it at some point? IIRC, migrating to Undici is still something we may want, so we're adding an additional road block here.

Reply:
elastic-transport-js's `UndiciConnection` does [not currently accept](https://github.com/elastic/elastic-transport-js/blob/main/src/connection/UndiciConnection.ts#L62,L64) `agentFn` as input configuration, aka it does not accept factories.

Thus, I couldn't simply pass in the factories provided by `AgentManager`.
Since we're still using standard `HttpConnection` ATM I didn't want to go too far with my PR.

That's why I made `agentManager` parameter optional too. If we want to switch to `undici` we can simply stop passing an agent manager, and undici's default logic for creating its agent will apply.

I haven't gotten into the internals of unidici, but if we then consider appropriate to use agent factories with undici as well, we'll have to:

1. Support factories in elastic-transport-js's `UndiciConnection` class.
2. Update the `AgentManager` to support undici mode as well.

Reply:
I think this issue covers everything we want https://github.com/elastic/elasticsearch-js/issues/1740

Ideally we don't want to have to create this factory / single agent workaround, the client should allow us to do that. So I hope when we move to undici that we'd just remove the AgentManager altogether and get everything out of the box 🌈